File: 2021/c/smaller-strings/v1.py
def printresult(order, result):
    print(f"Case #{order + 1}: {result}")


def readln():
    return input().strip()


def readint():
    return int(input().strip())


def readints():
    return list(map(int, input().split()))


def hashc(c):
    return ord(c) - 96


MOD = int(10e9 + 7)


def power(base, exp):
    """Fast power calculation using repeated squaring"""
    ans = 1
    while exp:
        if exp & 1:
            ans = (ans * base) % MOD
        exp >>= 1
        base *= base
    return ans % MOD


def solve(order):
    N, K = readints()
    S = readln()
    if N == 1:
        printresult(order, hashc(S) - 1)
        return
    if N == 2:
        f, s = hashc(S[0]), hashc(S[1])
        printresult(order, min(f, s))
        return

    total = 0
    mid = N // 2
    if N % 2 == 0:
        mid -= 1
    all_good = True
    for i in reversed(range(mid)):
        if S[i] < S[N - i - 1]:
            all_good = False
            break
    if N % 2 == 0:
        total = hashc(S[mid]) if S[mid] < S[N - mid - 1] else hashc(S[N - mid - 1])
        print(S[mid], hashc(S[mid]))
        if not all_good:
            total -= 1
        print(total)
        for i in reversed(range(mid)):
            total = ((hashc(S[i]) - 1) * power(K, mid - i) + total) % MOD
    else:
        pass

    printresult(order, total)


if __name__ == "__main__":
    T = readint()
    for t in range(T):
        solve(t)

File: 2021/c/smaller-strings/v1.cpp
#include <bits/stdc++.h>
using namespace std;

typedef vector<int> vt;
typedef vector<vector<int>> vvt;
typedef vector<string> vs;

#define repeat(ix, T) for (int ix = 0; ix < T; ++ix)
#define FOR(ix, s, e) for (int ix = s; ix < e; ++ix)
#define FORR(ix, s, e) for (int ix = s; ix > e; --ix)
#define map(ar, dr, f)                  \
    for (int i = 0; i < dr.size(); ++i) \
        dr[i] = f(ar[i]);
#define readint(s) \
    int s;         \
    cin >> s;
#define readstr(s) \
    string s;      \
    getline(cin >> ws, s);
#define readints(ar, S)         \
    auto ar = vt();             \
    for (int i = 0; i < S; ++i) \
    {                           \
        int e;                  \
        cin >> e;               \
        ar.push_back(e);        \
    }
#define prs(result) cout << "Case #" << order << ": " << result << endl;
template <class T>
void printarr(const vector<T> &ar)
{
    for (auto iter = ar.begin(); iter < ar.end() - 1; ++iter)
    {
        cout << *iter << " ";
    }
    cout << *(ar.end() - 1);
}

int MOD = int(1e9 + 7);

int num(char c)
{
    return c - 96;
}

void solve(int order)
{
    readint(N);
    readint(K);
    readstr(S);
    if (N == 1)
    {
        prs(num(S[0]) - 1);
        return;
    }
    if (N == 2)
    {
        if (S[0] < S[1])
        {
            prs(num(S[0]));
        }
        else
        {
            prs(num(S[0]) - 1);
        }
        return;
    }

    int mid = N / 2 - !(N & 1);
    long long total = 0;
    if (N & 1)
    {
        bool allGood = true;
        FORR(i, mid - 1, -1)
        {
            int opi = N - 1 - i;
            if (S[i] >= S[opi])
            {
                allGood = false;
                break;
            }
        }
        total = num(S[mid]) - (allGood ? 0 : 1);
    }
    else
    {

        int opi = N - 1 - mid;
        int curr = num(S[mid]) - ((S[mid] < S[opi]) ? 0 : 1);
        total = curr;
    }

    FORR(i, mid - 1, -1)
    {
        int opi = N - 1 - i;
        int curr = num(S[i]) - 1;
        long long coef = 1;
        if (curr > 0)
        {
            FOR(j, 0, mid - i)
            {
                coef = (K * coef) % MOD;
            }
        }
        total = (coef * curr % MOD + total) % MOD;
    }

    prs(total);
}

int main()
{
    readint(T);
    repeat(t, T) solve(t + 1);
    return 0;
}
